OpenRA.Mods.Common/Traits/Health.cs

```diff
@@ -17,13 +19,40 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
 	{
 		[Desc("HitPoints")]
 		public readonly int HP = 0;
-
-		[Desc("Physical size of the unit used for damage calculations. Impacts within this radius apply full damage.")]
-		public readonly WDist Radius = new WDist(426);
-
 		[Desc("Trigger interfaces such as AnnounceOnKill?")]
 		public readonly bool NotifyAppliedDamage = true;
 
+		[FieldLoader.LoadUsing("LoadShape")]
+		public readonly IHitShape Shape;
+
+		static object LoadShape(MiniYaml yaml)
+		{
+			string shape = "Circle";
+			int radius = 426;
+			int2[] verts = null;
+
+			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
+			if (shapeNode != null)
+			{
+				shape = shapeNode.Value.Value;
+
+				var nodes = shapeNode.Value.ToDictionary();
+
+				if (nodes.ContainsKey("Radius"))
+					radius = FieldLoader.GetValue<int>("Radius", nodes["Radius"].Value);
+
+				if (nodes.ContainsKey("Vertices"))
+					verts = FieldLoader.GetValue<int2[]>("Vertices", nodes["Vertices"].Value);
+			}
+
+			switch (shape)
+			{
+				case "Rectangle": return new Rectangle(verts[0], verts[1]);
```

Comment:
If you use reflection to create these and have the constructor accept a list of miniyaml nodes then you can avoid hardcoding the types, and modders could add new ones.